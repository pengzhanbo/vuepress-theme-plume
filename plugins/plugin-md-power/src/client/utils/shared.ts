const cache: {
  [src: string]: Promise<void> | undefined
} = {}
export function loadScript(src: string) {
  if (__VUEPRESS_SSR__)
    return Promise.resolve()

  if (document.querySelector(`script[src="${src}"]`)) {
    if (cache[src])
      return cache[src]
    return Promise.resolve()
  }

  const script = document.createElement('script')
  script.src = src
  document.body.appendChild(script)

  cache[src] = new Promise((resolve, reject) => {
    script.onload = () => {
      resolve()
      delete cache[src]
    }
    script.onerror = reject
  })
  return cache[src]
}

export function loadStyle(href: string, target: ShadowRoot) {
  if (__VUEPRESS_SSR__)
    return

  if (target.querySelector(`link[href="${href}"]`))
    return

  const link = document.createElement('link')
  link.rel = 'stylesheet'
  link.href = href
  target.appendChild(link)
}
