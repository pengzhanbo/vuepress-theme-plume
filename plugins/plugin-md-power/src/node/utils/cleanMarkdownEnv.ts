import type { MarkdownEnv } from 'vuepress/markdown'

export interface CleanMarkdownEnv extends MarkdownEnv {
  references?: unknown
  abbreviations?: unknown
  annotations?: unknown
}

export function cleanMarkdownEnv(env: CleanMarkdownEnv): CleanMarkdownEnv {
  return {
    base: env.base,
    filePath: env.filePath,
    filePathRelative: env.filePathRelative,
    references: env.references,
    abbreviations: env.abbreviations,
    annotations: env.annotations,
  }
}
