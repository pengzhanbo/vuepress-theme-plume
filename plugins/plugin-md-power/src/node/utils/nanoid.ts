import { customAlphabet } from 'nanoid'

export const nanoid: (size?: number) => string = customAlphabet('abcdefghijklmnopqrstuvwxyz', 5)
