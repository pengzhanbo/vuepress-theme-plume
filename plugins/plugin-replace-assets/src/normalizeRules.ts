import type { ReplaceAssetsPluginOptions, ReplacementRule } from './options.js'
import { isArray, isFunction } from '@vuepress/helper'
import { KNOWN_ASSET_EXTENSIONS, KNOWN_IMAGE_EXTENSIONS, KNOWN_MEDIA_EXTENSIONS } from './constants.js'

export function createFindPattern(extensions: string[]): RegExp {
  return new RegExp(`\\.(?:${extensions.join('|')})(\\?.*)?$`)
}

export function normalizeRules(options: ReplaceAssetsPluginOptions): ReplacementRule[] {
  const normalized: ReplacementRule[] = []

  if ((typeof options === 'string' || isFunction(options))) {
    options && normalized.push({
      find: createFindPattern(KNOWN_ASSET_EXTENSIONS),
      replacement: options,
    })
    return normalized
  }

  if (isArray(options)) {
    normalized.push(...options)
    return normalized
  }

  if ('find' in options) {
    options.find && options.replacement && normalized.push(options)
    return normalized
  }

  if (options.image) {
    normalized.push({
      find: createFindPattern(KNOWN_IMAGE_EXTENSIONS),
      replacement: options.image,
    })
  }

  if (options.media) {
    normalized.push({
      find: createFindPattern(KNOWN_MEDIA_EXTENSIONS),
      replacement: options.media,
    })
  }

  if (options.all) {
    normalized.push({
      find: createFindPattern(KNOWN_ASSET_EXTENSIONS),
      replacement: options.all,
    })
  }

  if (options.rules) {
    normalized.push(...isArray(options.rules) ? options.rules : [options.rules])
  }

  return normalized
}
