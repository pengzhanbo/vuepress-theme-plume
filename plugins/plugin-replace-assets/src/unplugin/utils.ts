import { removeEndingSlash, removeLeadingSlash } from '@vuepress/helper'

export function createAssetPattern(prefix: string) {
  const s = `(${prefix}.*?)`
  return new RegExp(
    [
      `(?:"${s}")`, // "prefix"
      `(?:'${s}')`, // 'prefix'
      `(?:\\(${s}\\))`, // (prefix)
      `(?:\\('${s}'\\))`, // ('prefix')
      `(?:\\("${s}"\\))`, // ("prefix")
      `(?:\\\\"${s}\\\\")`, // \"prefix\"
    ].join('|'),
    'gu',
  )
}

const htmlLangRE = /\.(?:html|htm)$/

export const isHTMLRequest = (request: string) => htmlLangRE.test(request)

const nonJsRe = /\.json(?:$|\?)/

export function isNonJsRequest(request: string): boolean {
  return nonJsRe.test(request)
}

export function normalizeUrl(url: string, base?: string): string {
  if (!url)
    return ''

  if (base) {
    url = `${removeEndingSlash(base)}/${removeLeadingSlash(url)}`
  }
  return url
}
