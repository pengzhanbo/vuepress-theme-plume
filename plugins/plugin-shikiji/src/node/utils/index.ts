import { Logger } from '@vuepress/helper'

export const logger = new Logger('@vuepress-plume/plugin-shikiji')

export * from './attrsToLines.js'
export * from './collapsedLines.js'
export * from './lru.js'
export * from './resolveAttr.js'
export * from './resolveLanguage.js'
export * from './whitespace.js'
