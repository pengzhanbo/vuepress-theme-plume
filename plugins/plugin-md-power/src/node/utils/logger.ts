/* istanbul ignore file -- @preserve */
/* eslint-disable no-console */
import { colors, ora } from 'vuepress/utils'

type Ora = ReturnType<typeof ora>

/**
 * Logger utils
 */
export class Logger {
  public constructor(
    /**
     * Plugin/Theme name
     */
    private readonly name = '',
  ) {}

  private init(subname: string, text: string): Ora {
    return ora({
      prefixText: colors.blue(`${this.name}${subname ? `:${subname}` : ''}: `),
      text,
    })
  }

  /**
   * Create a loading spinner with text
   */
  public load(subname: string, msg: string): {
    succeed: (text?: string) => void
    fail: (text?: string) => void
  } {
    const instance = this.init(subname, msg)

    return {
      succeed: (text?: string) => instance.succeed(text),
      fail: (text?: string) => instance.succeed(text),
    }
  }

  public info(subname: string, text = '', ...args: unknown[]): void {
    this.init(subname, colors.blue(text)).info()

    if (args.length)
      console.info(...args)
  }

  /**
   * Log success msg
   */
  public succeed(subname: string, text = '', ...args: unknown[]): void {
    this.init(subname, colors.green(text)).succeed()

    if (args.length)
      console.log(...args)
  }

  /**
   * Log warning msg
   */
  public warn(subname: string, text = '', ...args: unknown[]): void {
    this.init(subname, colors.yellow(text)).warn()

    if (args.length)
      console.warn(...args)
  }

  /**
   * Log error msg
   */
  public error(subname: string, text = '', ...args: unknown[]): void {
    this.init(subname, colors.red(text)).fail()

    if (args.length)
      console.error(...args)
  }
}

export const logger: Logger = new Logger('vuepress-plugin-md-power')
