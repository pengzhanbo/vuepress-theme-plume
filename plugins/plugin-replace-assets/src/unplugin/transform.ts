import type { ReplacementRule } from '../options.js'
import MagicString from 'magic-string'
import { normalizeUrl } from './utils.js'

const cache = new Map<string, string>()

export function transformAssets(code: string, pattern: RegExp, rules: ReplacementRule[]): string {
  const s = new MagicString(code)
  let matched: RegExpExecArray | null
  let hasMatched = false
  // eslint-disable-next-line no-cond-assign
  while ((matched = pattern.exec(code))) {
    const assetUrl = matched[6] || matched[5] || matched[4] || matched[3] || matched[2] || matched[1]
    const [left, right] = matched[0].startsWith('(')
      ? ['("', '")']
      : matched[0].startsWith('\\"')
        ? ['\\"', '\\"']
        : ['"', '"']

    const start = matched.index
    const end = start + matched[0].length
    const resolved = replacementAssetWithRules(rules, assetUrl)
    if (resolved) {
      hasMatched = true
      s.update(start, end, `${left}${resolved}${right}`)
    }
  }

  if (!hasMatched)
    return code

  return s.toString()
}

export function replacementAssetWithRules(rules: ReplacementRule[], url: string): string | void {
  if (cache.has(url))
    return cache.get(url)

  for (const { find, replacement } of rules) {
    if (find && isMatchUrl(find, url)) {
      let replaced = ''
      if (typeof replacement === 'function') {
        replaced = normalizeUrl(replacement(url))
      }
      else {
        replaced = normalizeUrl(url, replacement)
      }
      /* istanbul ignore if -- @preserve */
      if (replaced) {
        cache.set(url, replaced)
        return replaced
      }
    }
  }
  return undefined
}

export function isMatchUrl(find: string | RegExp, url: string): boolean {
  if (typeof find === 'string') {
    if (find[0] === '^' || find[find.length - 1] === '$') {
      return new RegExp(find).test(url)
    }
    else {
      return url.endsWith(find) || url.startsWith(find)
    }
  }

  return find.test(url)
}
