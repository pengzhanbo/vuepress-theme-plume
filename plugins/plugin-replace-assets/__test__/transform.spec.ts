import { describe, expect, it, vi } from 'vitest'
import { normalizeRules } from '../src/normalizeRules.js'
import { isMatchUrl, replacementAssetWithRules, transformAssets } from '../src/unplugin/transform.js'
import { createAssetPattern } from '../src/unplugin/utils.js'

describe('plugin-replace-assets > isMatchUrl', () => {
  it.each([
    {
      name: 'string like regexp with ^ and $',
      find: '^/images/.*\.(jpe?g|png|gif|svg)(\\?.*)?$',
      expects: [
        ['/images/foo.jpg', true],
        ['/images/foo.png', true],
        ['/images/foo.gif', true],
        ['/images/foo.svg', true],
        ['/images/foo.jpg?a=1', true],
        ['/images/foo.txt', false],
        ['/medias/foo.mp4', false],
      ] as const,
    },
    {
      name: 'string like regexp start with ^',
      find: '^/medias/',
      expects: [
        ['/medias/foo.mp4', true],
        ['/medias/foo.ogg', true],
        ['/medias/foo.ogv', true],
        ['/medias/foo.webm', true],
        ['/images/foo.jpg', false],
      ] as const,
    },
    {
      name: 'string like regexp end with $',
      find: '\.(jpe?g|png|gif|svg)$',
      expects: [
        ['/images/foo.jpg', true],
        ['/images/foo.png', true],
        ['/images/foo.gif', true],
        ['/images/foo.svg', true],
        ['/images/foo.txt', false],
        ['/medias/foo.mp4', false],
      ] as const,
    },
    {
      name: 'string start width pathname',
      find: '/images/',
      expects: [
        ['/images/foo.jpg', true],
        ['/images/foo.png', true],
        ['/images/foo.gif', true],
        ['/images/foo', true],
        ['/medias/foo.mp4', false],
      ] as const,
    },
    {
      name: 'string end width extension',
      find: '.jpg',
      expects: [
        ['/images/foo.jpg', true],
        ['/images/foo.png', false],
        ['/images/foo.gif', false],
        ['/images/foo', false],
        ['/medias/foo.mp4', false],
      ] as const,
    },
    {
      name: 'regexp',
      find: /^\/images\/.*\.(jpe?g|png|gif|svg)$/,
      expects: [
        ['/images/foo.jpg', true],
        ['/images/foo.png', true],
        ['/images/foo.gif', true],
        ['/images/foo.svg', true],
        ['/images/foo.txt', false],
        ['/medias/foo.mp4', false],
      ] as const,
    },
  ])('$name', ({ find, expects }) => {
    for (const [url, expected] of expects) {
      expect(isMatchUrl(find, url)).toBe(expected)
    }
  })
})

describe('plugin-replace-assets > replacementAssetWithRules', () => {
  const IMAGE_SUPPORTED = ['jpg', 'jpeg', 'png', 'gif', 'webp', 'svg', 'avif']
  const MEDIA_SUPPORTED = ['mp4', 'webm', 'ogg', 'mp3', 'wav', 'flac', 'aac', 'm3u8', 'm3u', 'flv', 'pdf']
  const replacementFn = vi.fn(url => `https://example.com/assets${url}`)

  it.each([
    {
      name: 'string replacement',
      rules: normalizeRules('https://example.com/assets/'),
      expects: [
        // images
        ...IMAGE_SUPPORTED.map(ext => [
          `/images/foo.${ext}`,
          `https://example.com/assets/images/foo.${ext}`,
        ]),
        // media
        ...MEDIA_SUPPORTED.map(ext => [
          `/medias/foo.${ext}`,
          `https://example.com/assets/medias/foo.${ext}`,
        ]),
        // have query string
        ['/images/foo.jpg?a=1', 'https://example.com/assets/images/foo.jpg?a=1'],
        // cached images
        ['/images/foo.jpg', 'https://example.com/assets/images/foo.jpg'],
        // no supported
        ['/images/foo.txt', undefined],
        ['/medias/foo', undefined],
      ] as const,
    },
    {
      name: 'function replacement',
      rules: normalizeRules(replacementFn),
      expects: [
        // images
        ...IMAGE_SUPPORTED.map(ext => [
          `/images-1/foo.${ext}`,
          `https://example.com/assets/images-1/foo.${ext}`,
        ]),
        // media
        ...MEDIA_SUPPORTED.map(ext => [
          `/medias-1/foo.${ext}`,
          `https://example.com/assets/medias-1/foo.${ext}`,
        ]),
        // have query string
        ['/images-1/foo.jpg?a=1', 'https://example.com/assets/images-1/foo.jpg?a=1'],
        // cached images
        ['/images-1/foo.jpg', 'https://example.com/assets/images-1/foo.jpg'],
        // no supported
        ['/images-1/foo.txt', undefined],
        ['/medias-1/foo', undefined],
      ] as const,
    },
  ])('$name', ({ name, rules, expects }) => {
    for (const [url, expected] of expects) {
      expect(replacementAssetWithRules(rules, url)).toBe(expected)
    }
    if (name === 'function replacement') {
      // should not called with cached, and not called with no supported
      expect(replacementFn).toBeCalledTimes(expects.length - 3)
    }
  })
})

describe('plugin-replace-assets > transformAssets', () => {
  const rules = normalizeRules('https://example.com/assets/')
  const pattern = createAssetPattern('/[^/]')

  it('should work with like html', () => {
    const source = `\
<img src="/images/foo.jpg" />
<img src="/images/foo.png" />
<img src="/images/foo.gif" />
<img src="/images/foo.svg" />
<img src="/images/foo.txt" />
<img src="/medias/foo.mp4" />

<img src="/images/foo.jpg?a=1" />

<img src="https://not-replace.com/images/foo.jpg" />

<video src="/medias/foo.mp4" />
<audio src="/medias/foo.mp3" />

<embed src="/medias/foo.pdf" />
`
    expect(transformAssets(source, pattern, rules)).toMatchSnapshot()
  })

  it('should work with like css', () => {
    const source = `\
.foo {
  background-image: url("/images/foo.jpg");
  background-image: url("/images/foo.png");
  background-image: url("/images/foo.gif");
  background-image: url("/images/foo.svg");
  background-image: url("/medias/foo.mp4");

  background-image: url('/images/foo.jpg');
  background-image: url(/images/foo.png);

  background-image: url("/images/foo.jpg?a=1");

  background-image: url("https://not-replace.com/images/foo.jpg");

  background: url('/images/foo.png');

}
`
    expect(transformAssets(source, pattern, rules)).toMatchSnapshot()
  })

  it('should work with like js', () => {
    const source = `\
  const a = "/images/foo.jpg"
  const b = '/images/foo.jpg'
  const c = '/images/foo.jpg?a=1'
  const d = "https://not-replace.com/images/foo.jpg"

  const json_string = JSON.parse("{\\"a\\":\\"/images/foo.jpg\\"}")
`

    expect(transformAssets(source, pattern, rules)).toMatchSnapshot()
  })

  it('should work with no match', () => {
    const source = `\
<video src="./medias/foo.mp4" />
<img src="https://not-replace.com/images/foo.jpg" />

const a = "images/foo.jpg"
`
    expect(transformAssets(source, pattern, rules)).toBe(source)
  })
})
