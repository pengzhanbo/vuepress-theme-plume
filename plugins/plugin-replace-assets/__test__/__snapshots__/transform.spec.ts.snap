// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-replace-assets > transformAssets > should work with like css 1`] = `
".foo {
  background-image: url("https://example.com/assets/images/foo.jpg");
  background-image: url("https://example.com/assets/images/foo.png");
  background-image: url("https://example.com/assets/images/foo.gif");
  background-image: url("https://example.com/assets/images/foo.svg");
  background-image: url("https://example.com/assets/medias/foo.mp4");

  background-image: url("https://example.com/assets/images/foo.jpg");
  background-image: url("https://example.com/assets/images/foo.png");

  background-image: url("https://example.com/assets/images/foo.jpg?a=1");

  background-image: url("https://not-replace.com/images/foo.jpg");

  background: url("https://example.com/assets/images/foo.png");

}
"
`;

exports[`plugin-replace-assets > transformAssets > should work with like html 1`] = `
"<img src="https://example.com/assets/images/foo.jpg" />
<img src="https://example.com/assets/images/foo.png" />
<img src="https://example.com/assets/images/foo.gif" />
<img src="https://example.com/assets/images/foo.svg" />
<img src="/images/foo.txt" />
<img src="https://example.com/assets/medias/foo.mp4" />

<img src="https://example.com/assets/images/foo.jpg?a=1" />

<img src="https://not-replace.com/images/foo.jpg" />

<video src="https://example.com/assets/medias/foo.mp4" />
<audio src="https://example.com/assets/medias/foo.mp3" />

<embed src="https://example.com/assets/medias/foo.pdf" />
"
`;

exports[`plugin-replace-assets > transformAssets > should work with like js 1`] = `
"  const a = "https://example.com/assets/images/foo.jpg"
  const b = "https://example.com/assets/images/foo.jpg"
  const c = "https://example.com/assets/images/foo.jpg?a=1"
  const d = "https://not-replace.com/images/foo.jpg"

  const json_string = JSON.parse("{\\"a\\":\\"https://example.com/assets/images/foo.jpg\\"}")
"
`;
